### PR TITLE
Simplify Container Status Patching

### LINKED ISSUE #410: Make patching container statuses more approachable
Right now, each provider has to manually send patch requests to kubernetes to update their container statuses. This seems ripe for refactoring to make it simpler for each provider to provide status updates.

https://github.com/deislabs/krustlet/blob/0e838ab046b32b70d5da855b59354867e7262f7e/crates/wasi-provider/src/states/running.rs#L45-L56

### PR DESCRIPTION
**Pull request by kflansburg**
Closes #410

### CODE REVIEW

File: crates/kubelet/src/container/status.rs
Comment (kflansburg) [inline crates/kubelet/src/container/status.rs:90]: This relies on container statuses matching the containers in `Pod.containers` and `Pod.init_containers`, in the same order. I believe that this would be best practice.

File: crates/kubelet/src/pod/mod.rs
Comment (kflansburg) [inline crates/kubelet/src/pod/mod.rs:169]: This seemed inconsistent with the other two methods `containers` and `init_containers`.

File: crates/kubelet/src/pod/status.rs
Comment (kflansburg) [inline crates/kubelet/src/pod/status.rs:8]: This initializes Pod status to include containers in the correct order as expected by `patch_container_status`. It is kind of required, so I'm wondering if this should be done prior to entering the first state handler.

File: crates/wasi-provider/src/states/running.rs
Comment (kflansburg) [inline crates/wasi-provider/src/states/running.rs:42]: I don't agree with `fail_fatal` here or in the init pod state, but it appears to be required by the integration tests. The tests expect the Pod to fail whereas I would expect it to go to `Error`. Incidentally there appears to be a bug when the Pod does go to `Error` and then back to `Registered`, the wasm runtime encounters some error with a channel. We may need to do some clean up in the error state in anticipation of going back to `Registered`.

File: crates/wasi-provider/src/states/starting.rs
Comment (kflansburg) [inline crates/wasi-provider/src/states/starting.rs:115]: This is somewhat tied to making all of these handlers infallible, but this one in particular looked like it would drop important container status patches. The same can be said of the `Initializing` state handler.

File: crates/wasi-provider/src/wasi_runtime.rs
Comment (kflansburg) [inline crates/wasi-provider/src/wasi_runtime.rs:302]: This was the one Err in this method that did not notify the channel.

**Review by kflansburg — COMMENTED** [review]
This was a first crack at simplifying container status patches. There may be some nits here, but I think this is supporting work for the state machine, so I would expect it to get merged and then the container state machine will likely involve further refactoring.

File: crates/kubelet/src/pod/status.rs
Comment (bacongobbler) [reply crates/kubelet/src/pod/status.rs:8]: As an aside, this comment would be great to add to the doc comments above on line 7.

File: crates/wasi-provider/src/states/running.rs
Comment (bacongobbler) [reply crates/wasi-provider/src/states/running.rs:42]: Would it be all right if we pulled this out into a separate PR? That way we can have this discussion separate from the container status patching logic.

File: crates/wasi-provider/src/wasi_runtime.rs
Comment (bacongobbler) [reply crates/wasi-provider/src/wasi_runtime.rs:302]: Good catch. I swore I updated this to the new notify logic...

**Comment by bacongobbler** [comment]
Looks great. Going to take this for a spin to make sure things work as expected. But the fact that the container status patching logic was removed from the wasi provider is huge!

File: crates/wasi-provider/src/states/running.rs
Comment (itowlson) [reply crates/wasi-provider/src/states/running.rs:42]: There was discussion about this on the 'remove panics from handlers' PR - this was to replicate existing behaviour which was relied on by the integration tests.  I think everyone agrees that it is not the desired behaviour, but changing the behaviour should be tackled as a specific work item (which will be non-trivial), not rolled in along with other work.

Comment (itowlson) [reply crates/wasi-provider/src/states/running.rs:42]: Never mind, I see you already raised an issue to do this as a separate item.  Thanks!

File: crates/kubelet/src/pod/mod.rs
Comment (itowlson) [reply crates/kubelet/src/pod/mod.rs:169]: I suspect at the time it was written it was used primarily for getting keys/IDs rather than the full container objects. But it looks like now the only places it is used would work just as well with the full container objects.  There is also a place in `store/mod.rs` that explicitly constructs its own `all_containers` collection which this modified method could supplant.

**Review by itowlson — CHANGES_REQUESTED** [review]
This will be great to have!  Unfortunately I have reservations about relying so absolutely on the order of container statuses in the status section being in sync with the order of specs in the spec section.  It seems like if anything causes this assumption to become violated we have no way to detect and adapt (or even fail informatively) - we will just blindly write statuses to the wrong container.  Are we that confident about the assumption, and if not can we come up with an alternative approach?

File: crates/kubelet/src/container/status.rs
Comment (kflansburg) [reply crates/kubelet/src/container/status.rs:95]: Did not realize `ContainerKey` included init vs. app. Will change this.

File: crates/kubelet/src/container/status.rs
Comment (kflansburg) [reply crates/kubelet/src/container/status.rs:105]: Makes sense.

File: crates/kubelet/src/container/status.rs
Comment (kflansburg) [reply crates/kubelet/src/container/status.rs:108]: There are four, I do not know if the resource version is actually necessary, but it was in the original WASI provider code and I added it back when I was debugging. In general I dont think its an issue to use a "magic number" in this case, as long as the number is actually correct. As far as the vec literal, that seems reasonable, although I will probably not create 3+ methods for that, just define the ReplaceOperation inside the vec literal.

File: crates/kubelet/src/container/status.rs
Comment (kflansburg) [reply crates/kubelet/src/container/status.rs:143]: Yes, I believe that syntax works.

File: crates/kubelet/src/container/status.rs
Comment (kflansburg) [reply crates/kubelet/src/container/status.rs:114]: I do not think it is guaranteed. I dont think it is reasonable to fetch the Pod each time it needs to be patched. We could store some up-to-date version of the Pod status, and overwrite the entire status when we patch a container. Is there any expectation that things other than than the scheduled kubelet update the status of the Pod under normal operation?

Comment (thomastaylor312) [reply crates/kubelet/src/container/status.rs:114]: FWIW, I tried to write this as a JSON patch the first time I was tackling it as I didn't like the idea of fetching the pod each time. This ordering problem was the stickler then too. However, I think we might be able to remove some of the overhead of fetching the Pod each time if we switch to having a reflector. A reflector implementation updates its cache each time an object updates, so then we can just fetch the object from the cache. We could still end up with a case where that object updates after we fetch from the cache, but that should be fairly unlikely as I don't think anything else but the master could update a pod status. What I don't know is how many controllers might try to update pod statuses though.

File: crates/kubelet/src/container/status.rs
Comment (thomastaylor312) [reply crates/kubelet/src/container/status.rs:143]: Yep, the syntax is

```rust
let thing = PatchParams {
    foo: "bar",
    ..Default::default()
};
```

**Comment by kflansburg** [comment]
I have addressed all comments, except for switching to a Reflector. What I think makes the most sense is that `Pod::queue` should update the Pod definition that it is driving the state machine with whenever updates arrive. This would effectively be a reflector, and mean that the `Pod` passed to `json_status` will be the latest definition. I think this is the simplest solution and avoids having some shared state of Pod definitions that `patch_container_status` would need access to.

File: crates/kubelet/src/state.rs
Comment (kflansburg) [inline crates/kubelet/src/state.rs:268]: I have implemented my suggestion of applying this initial status update prior to applying the Provider's first status. I also removed the use of this function in the Providers to verify that this eliminates their responsibility to do this.

**Review by kflansburg — COMMENTED** [review]
This last commit implements my suggestion of updating the Pod manifest that is supplied to the State handlers.

File: crates/kubelet/src/state.rs
Comment (thomastaylor312) [inline crates/kubelet/src/state.rs:269]: Nit: consider doing this in a block (along with the `read()` call) so as to avoid the need for a manual `drop`

**Review by thomastaylor312 — APPROVED** [review]
I think with the new method of doing this, I am now ok with merging. Let's wait on @itowlson for his final review

File: crates/kubelet/src/state.rs
Comment (thomastaylor312) [inline crates/kubelet/src/state.rs:288]: Ok, think I have a few concerns about this. This is mainly broken down into two things 1) I don't know if this is any better than our old method of doing things and 2) If we do continue doing it this way, I think there could be a better way of doing this. 

1. So I don't know if this provides any benefit on our previous version other than API call reduction. We still have a bunch of waiting and things that we end up doing (no matter how we implement the retry). I think it could actually end up being slower as things step through the state machine
2. If we do want to go this route, I think we should reuse/export the `retry!` macros from the node code at the very minimum. I think a more elegant solution could entail wrapping the pod in a custom struct with a `RwLock` that write locks until it receives a response from the API and then abstract that all away behind `pod.read()` like we already have

**Review by thomastaylor312 — CHANGES_REQUESTED** [review]
I have some concerns with the loop approach in the latest commit. Detailed in the comments below

File: crates/kubelet/src/state.rs
Comment (kflansburg) [reply crates/kubelet/src/state.rs:288]: So regarding performance, I think this is better for two reasons:
* this loop only runs when a Pod is first scheduled, not on each state change.
* in most cases, all it does is wait a second or two for the reflector to update, and we don't have to try to patch multiple times. 
I will address the other nits.

Comment (thomastaylor312) [reply crates/kubelet/src/state.rs:288]: > this loop only runs when a Pod is first scheduled, not on each state change.

Ok, so I misread that. I do think we should have a max number of retries (preferably as a `const`) like what you have now and then just error out and let k8s know the pod has errors (this might be a good use of the `Unknown` phase)

Comment (kflansburg) [reply crates/kubelet/src/state.rs:288]: Sounds good will do.

File: crates/kubelet/src/pod/mod.rs
Comment (kflansburg) [reply crates/kubelet/src/pod/mod.rs:162]: I dont think you can really do this: `.unwrap_or_else(|| &PodStatus::default())`, because the reference doesnt outlive the closure. I will see what I can do though.

**Review by thomastaylor312 — APPROVED** [review]

**Pull request merged**
